src/ess/reduce/time_of_flight/toa_to_tof.py

```diff
@@ -111,231 +104,289 @@ def compute_tof_lookup_table(
+        # a simple modulo also works here because even if they end up between 0 and
+        # 0.5*bin_width, we are (below) computing the mean between -0.5*bin_width and
+        # 0.5*bin_width and it yields the same result.
+        data.coords['event_time_offset'] %= pulse_period - time_bins_half_width
 
-def masked_tof_lookup_table(
-    tof_lookup: TimeOfFlightLookupTable,
-    error_threshold: LookupTableRelativeErrorThreshold,
-) -> MaskedTimeOfFlightLookupTable:
-    """
-    Mask regions of the lookup table where the variance of the projected time-of-flight
-    is larger than a given threshold.
+        binned = data.group('pulse').bin(
+            distance=dist_edges + simulation_distance, event_time_offset=time_bins
+        )
 
-    Parameters
-    ----------
-    tof_lookup:
-        Lookup table giving time-of-flight as a function of distance and
-        time-of-arrival.
-    variance_threshold:
-        Threshold for the variance of the projected time-of-flight above which regions
-        are masked.
-    """
-    relative_error = sc.stddevs(tof_lookup.data) / sc.values(tof_lookup.data)
+        # Weighted mean of tof inside each bin
+        mean_tof = (
+            binned.bins.data * binned.bins.coords["tof"]
+        ).bins.sum() / binned.bins.sum()
+        # Compute the variance of the tofs to track regions with large uncertainty
+        variance = (
+            binned.bins.data * (binned.bins.coords["tof"] - mean_tof) ** 2
+        ).bins.sum() / binned.bins.sum()
+
+        mean_tof.variances = variance.values
+        mean_tof.coords["distance"] = sc.midpoints(mean_tof.coords["distance"])
+        pieces.append(mean_tof)
+
+    table = sc.concat(pieces, 'distance')
+    table.coords["event_time_offset"] = sc.midpoints(table.coords["event_time_offset"])
+
+    # We are still missing the upper edge of the table in the event_time_offset axis
+    # (at pulse_period). Because the event_time_offset is periodic, we can simply copy
+    # the left edge over to the right edge.
+    # Note that this needs to be done pulse by pulse, as the left edge of the second
+    # pulse is the same as the right edge of the first pulse, and so on (in the case
+    # of pulse_stride > 1).
+
+    # First, extend the table to the right by 1, and set the coordinate to pulse_period.
+    slab = sc.empty_like(table['event_time_offset', 0])
+    slab.coords['event_time_offset'] = pulse_period
+    table = sc.concat([table, slab], dim='event_time_offset')
+    # Then, copy over the values. Instead of using pulse_stride, we use the number of
+    # pulses in the table, as it could be that there were no events in the first pulse.
+    npulses = table.sizes['pulse']
+    for i in range(npulses):
+        pulse = (i + 1) % npulses
+        left_edge = table.data['pulse', pulse]['event_time_offset', 0]
+        table.data['pulse', i]['event_time_offset', -1] = left_edge
```

Comment:
While this does not look wrong, I am a bit confused as to why we need to handle pulse-skipping as done above. Why group by pulse early and then deal with the mess here? Can't we just process based on the actual frame length (say 2*71ms), and then `fold` to obtain a pulse dim in the end?

Not saying it has to change, just wondering if it is simpler, without having thought through every detail.

Reply:
After looking at this for too long, I get lost in the details and don't always manage to take a step back and think about it more on the higher conceptual level.

This is a great suggestion, I implemented it and it works :+1: